### PR TITLE
kvm: honor migrate.wait and abort vm migration job

### LINKED ISSUE #3742: migratewait does not seem to be honored in KVM
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
API
~~~

##### CLOUDSTACK VERSION

~~~
4.14 and less
~~~

##### SUMMARY

While observing the logs and running migrations noticed:
job: job-1843/job-1915 initiate migrate command at 2019-12-03 13:14:29,833

later on at 2019-12-03 13:14:54,097 it is being reported that the job is completed this is about 30 seconds. 

meanwhile the migratewait setting is being set to 1, and seems to be honored at Xen 

with VMware we have the same failure: 

migrte command started at: 2019-12-03 13:21:14,585

and reported success at:  2019-12-03 13:21:24,044

About 10 seconds. 

##### STEPS TO REPRODUCE

~~~
Set the migratewait setting to 1 and initiate a livemigration
~~~

##### EXPECTED RESULTS

~~~
migratewait to be honored 
~~~

##### ACTUAL RESULTS

~~~
It's not
~~~

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR fixes #3742 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan test ubuntu20 kvm-ubuntu20

**Comment by weizhouapache** [comment]
@blueorangutan test

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java [hunk 1]
```diff
@@ -219,6 +220,29 @@ Use VIR_DOMAIN_XML_SECURE (value = 1) prior to v1.0.0.
...
+                        state = dm.getInfo().state;
+                    } catch (final LibvirtException e) {
+                        s_logger.info("Couldn't get VM domain state after " + sleeptime + "ms: " + e.getMessage());
+                    }
+                    if (state != null && state == DomainState.VIR_DOMAIN_RUNNING) {
+                        try {
+                            DomainJobInfo job = dm.getJobInfo();
+                            s_logger.info("Aborting domain job: " + job);
+                            dm.abortJob();
+                            result = "Migration was cancelled by cloudstack due to time out";
```
Comment (sureshanaparti) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java:237]: vmName (and time out), might be useful in the log here

**Comment by weizhouapache** [comment]
update pr as per @sureshanaparti 's comment. cc @nvazquez 

agent.log
```
2021-09-02 07:58:21,056 INFO  [resource.wrapper.LibvirtMigrateCommandWrapper] (agentRequest-Handler-5:null) (logid:09947a36) Aborting i-2-181-VM domain job: type:2
timeElapsed:1100
timeRemaining:0
dataTotal:542187520
dataProcessed:170979005
dataRemaining:254967808
memTotal:542187520
memProcessed:170979005
memRemaining:254967808
fileTotal:0
fileProcessed:0
fileRemaining:0

2021-09-02 07:58:21,061 DEBUG [resource.wrapper.LibvirtMigrateCommandWrapper] (agentRequest-Handler-5:null) (logid:09947a36) Migration of VM i-2-181-VM was cancelled by cloudstack due to time out after 1 seconds
2021-09-02 07:58:21,062 INFO  [resource.wrapper.LibvirtMigrateCommandWrapper] (agentRequest-Handler-5:null) (logid:09947a36) Migration thread for i-2-181-VM is done
```


management-server.log
```
2021-09-02 07:58:21,921 ERROR [c.c.a.ApiAsyncJobDispatcher] (API-Job-Executor-4:ctx-17d048fb job-2372) (logid:09947a36) Unexpected exception while executing org.apache.cloudstack.api.command.admin.vm.MigrateVMCmd
com.cloud.utils.exception.CloudRuntimeException: Exception during migrate: org.libvirt.LibvirtException: operation aborted: migration out job: canceled by client
	at com.cloud.vm.VirtualMachineManagerImpl.migrate(VirtualMachineManagerImpl.java:2772)

```

**Review by nvazquez — APPROVED** [review]
LGTM thanks @weizhouapache

**Review by sureshanaparti — APPROVED** [review]
code LGTM

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
Ping @borisstoyanov @vladimirpetrov can you verify this fix?

**Comment by nvazquez** [comment]
Hi @NuxRo kindly update/review after your tests

**Comment by NuxRo** [comment]
@nvazquez did some preliminary tests using the prebuilt packages for this PR and they failed, ie migration was not aborted, but I want to rerun them, make sure I haven't missed anything.

**Comment by weizhouapache** [comment]
> @nvazquez did some preliminary tests using the prebuilt packages for this PR and they failed, ie migration was not aborted, but I want to rerun them, make sure I haven't missed anything.

@NuxRo what's the value in global setting ?  some small vms can be migrated successfully in very few seconds.

**Comment by NuxRo** [comment]
@nvazquez @weizhouapache guys this works as expected, I was probably doing something dodgy the first time.
One thing though, the UI displayes this error:
"Exception during migrate: org.libvirt.LibvirtException: operation aborted: migration out job: cancelled by client"

Any chance we can make this more meaningful to the user, something involving "migratewait timeout reached, check your settings" ?

**Comment by weizhouapache** [comment]
> @nvazquez @weizhouapache guys this works as expected, I was probably doing something dodgy the first time.
> One thing though, the UI displayes this error:
> "Exception during migrate: org.libvirt.LibvirtException: operation aborted: migration out job: cancelled by client"
> 
> Any chance we can make this more meaningful to the user, something involving "migratewait timeout reached, check your settings" ?

@NuxRo good, thanks for testing !
I will update this pr.

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Pull request merged**